### PR TITLE
Add fitting capability to Eng Diff UI fitting tab

### LINKED ISSUE #27780: Engineering Fitting Tab - Running Fits
### Story:
As a user, once I have selected the runs that I wish to plot, I want to be able to select peak ranges and run the PseudoVoigt + BackToBackExponential fitting algorithms on each of the selected peaks, seeing the output.

### Details:
- [ ] Plot should have the option to convert to d-spacing using the DIFA, DIFC, and ZERO values generated from the Calibrate tab

- [x] The user should be able to select the peaks they wish to fit using the plotting widget. 

- [x] The result of the fitting algorithms should be overlaid over the top of the data points. 
- [ ] The results of the fitting should be output and stored for saving later.

### Tests:
- Fit results are saved somehow
- Both fit functions are run on each peak. 
- Two close peaks are still able to be fitted.

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**
- Adds the fit property browser to toolbar.
 - Allows for subtraction of background (so can fit all spectra in a single fit) - for which a new algorithm was made EnggEstimateFocussedBackground.
- Can now load in data as d-spacing or TOF

**To test:**
(1) Perform calibration and focusing as per instructions in PR #28393  
(or download this file [Uploading ENGINX_305761_bank_1.zip…])
(2) Toggle the fit property browser using the 'Fit' button on the toolbar.
(3) Add some functions and fit
(4) Play with background parameters (column 4 onward in table) before and after the subtract background is set to True (it should update the plot if the background if set to True)
(5) Select a row for which subtract background is set to True and click the `Inspect Background' button - it should pop up an external figure which will also update as the parameters are changed.

Partially fixes #27780

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — CHANGES_REQUESTED** [review]
Attempting to launch mantid results in this error. I think it may be to do with loading matplotlib before the backend is properly selected. 
```
Traceback (most recent call last):
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/app/start.py", line 192, in start
    exit_value = start_workbench(app, options)
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/app/start.py", line 127, in start_workbench
    main_window.setup()
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/app/mainwindow.py", line 163, in setup
    self.ipythonconsole = JupyterConsole(self)
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plugins/jupyterconsole.py", line 46, in __init__
    font_family=font.family(), font_size=font.pointSize())
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/jupyterconsole.py", line 60, in __init__
    shell.ex(startup_code)
  File "/usr/lib/python3/dist-packages/IPython/core/interactiveshell.py", line 2423, in ex
    exec(cmd, self.user_global_ns, self.user_ns)
  File "<string>", line 3, in <module>
  File "/usr/lib/python3/dist-packages/matplotlib/pyplot.py", line 116, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/usr/lib/python3/dist-packages/matplotlib/backends/__init__.py", line 60, in pylab_setup
    [backend_name], 0)
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plotting/backend_workbench.py", line 28, in <module>
    from workbench.plotting.figuremanager import (QAppThreadCall, new_figure_manager as new_figure_manager_impl,
  File "/home/conor/repos/mantid-review/qt/applications/workbench/workbench/plotting/figuremanager.py", line 31, in <module>
    from mantidqt.widgets.fitpropertybrowser import FitPropertyBrowser
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/fitpropertybrowser/__init__.py", line 12, in <module>
    from .fitpropertybrowser import FitPropertyBrowserBase, FitPropertyBrowser
  File "/home/conor/repos/mantid-review/qt/python/mantidqt/widgets/fitpropertybrowser/fitpropertybrowser.py", line 16, in <module>
    import matplotlib.pyplot as plt
AttributeError: module 'matplotlib' has no attribute 'pyplot'
```

File: scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_toolbar.py [hunk 1]
```diff
@@ -16,22 +16,28 @@
...
+        ('Fit', 'Open/close fitting tab', None, 'toggle_fit', False),
     )
 
     def _init_toolbar(self):
         for text, tooltip_text, mdi_icon, callback, checked in self.toolitems:
             if text is None:
                 self.addSeparator()
             else:
-                action = self.addAction(get_icon(mdi_icon), text, getattr(self, callback))
+                # action = self.addAction(get_icon(mdi_icon), text, getattr(self, callback))
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_toolbar.py:36]: good spot removed

File: Framework/PythonInterface/plugins/algorithms/EnggEstimateFocussedBackground.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        for ispec in range(0, nspec):
+            prog_reporter.report()
+
+            # get n points in convolution window
+            binwidth = np.mean(np.diff(inws.readX(ispec)))
+            nwindow = int(np.ceil(xwindow / binwidth))
+            if not nwindow % 2:
+                nwindow += 1
+
+            if nwindow < 3:
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/EnggEstimateFocussedBackground.py:90]: Not sure what you mean here. The xwindow has to be odd, greater than three and less than the number of of bins. I could combine these conditions as so?
```
if 3  <= nwindow < nbins:
	if not nwindow % 2:
		nwindow += 1
else
	# not effective due to edge effects of the convolution
	raise RuntimeError(f"The convolution window ({nwindow} points) must have at least three points but less than the length of the data ({nbins} points)")
'''

[the remark this answers is not in the document]
Comment (ConorMFinn) [reply Framework/PythonInterface/plugins/algorithms/EnggEstimateFocussedBackground.py:90]: So my reasoning was that `3` appears to be a magic number and is used in a couple of places to say there must be this many points. My thinking is that it should be moved to a variable like `MINIMUM_POINTS` or something so you can add a comment saying why it needs to be 3.

**Review by cailafinn — CHANGES_REQUESTED** [review]
Code is looking good to me now. Did notice during testing that repeadedly clicking the SubtractBG checkbox appears to be either applying or removing the subtraction multiple times. The resulting plot begins to look fairly warped after a few clicks.

**Comment by RichardWaiteSTFC** [comment]
> Code is looking good to me now. Did notice during testing that repeadedly clicking the SubtractBG checkbox appears to be either applying or removing the subtraction multiple times. The resulting plot begins to look fairly warped after a few clicks.

Yep good spot. seems to be a bug that got introduced when I refactored to put stuff in model. For some reason the background subtraction box doesn't uncheck but it does trigger the background to be added back...I'm looking into it.

**Review by cailafinn — CHANGES_REQUESTED** [review]
Code looks good and the new fitting and background functionality works really well. 

Should've spotted this earlier, but could you add some release notes and add the new functionality to the Engineering Diffraction documentation?

**Review by cailafinn — APPROVED** [review]
Thanks for adding those docs. This looks good to me now. Awesome work 👍

**Comment by RichardWaiteSTFC** [comment]
> Thanks for adding those docs. This looks good to me now. Awesome work 👍

Thanks for reviewing it - turned out to be a bit of a slog!

**Pull request merged**